// @anorcle/dsa/dist/cjs/Queue.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const InvalidOperationError_js_1 = __importDefault(require("./errors/InvalidOperationError.js"));
const index_js_1 = require("./index.js");
class Queue {
    constructor(initialArray = []) {
        this.list = new index_js_1.LinkedList(initialArray);
    }
    get size() {
        return this.list.size;
    }
    get empty() {
        return this.list.size == 0;
    }
    get front() {
        if (this.list.size == 0) {
            throw new InvalidOperationError_js_1.default('Unable to access element from Empty Queue');
        }
        return this.list.front;
    }
    push(value) {
        this.list.pushBack(value);
        return this.list.size;
    }
    pop() {
        if (this.list.size === 0) {
            throw new InvalidOperationError_js_1.default('Unable to pop element from empty queue');
        }
        return this.list.popFront();
    }
}
exports.default = Queue;
