// @anorcle/dsa/dist/cjs/Stack.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const InvalidOperationError_js_1 = __importDefault(require("./errors/InvalidOperationError.js"));
class Stack {
    constructor(initialArray = []) {
        this.array = initialArray;
    }
    get empty() {
        return this.array.length == 0;
    }
    get top() {
        const value = this.array.at(-1);
        if (value == undefined) {
            throw new InvalidOperationError_js_1.default('Unable to access element from empty stack');
        }
        else {
            return value;
        }
    }
    push(element) {
        return this.array.push(element);
    }
    pop() {
        const value = this.array.pop();
        if (value == undefined) {
            throw new InvalidOperationError_js_1.default('Unable to pop element from empty stack');
        }
        else {
            return value;
        }
    }
}
exports.default = Stack;
