// @anorcle/dsa/dist/mjs/Deque.js
import InvalidOperationError from './errors/InvalidOperationError.js';
import { LinkedList } from './index.js';
class Deque {
    list;
    constructor(initialArray = []) {
        this.list = new LinkedList(initialArray);
    }
    get size() {
        return this.list.size;
    }
    get empty() {
        return this.list.size == 0;
    }
    get front() {
        if (this.list.size == 0) {
            throw new InvalidOperationError('Unable to access element from Empty Deque');
        }
        return this.list.front;
    }
    get back() {
        if (this.list.size == 0) {
            throw new InvalidOperationError('Unable to access element from Empty Deque');
        }
        return this.list.back;
    }
    pushBack(value) {
        this.list.pushBack(value);
        return this.list.size;
    }
    pushFront(value) {
        this.list.pushFront(value);
        return this.list.size;
    }
    popBack() {
        if (this.list.size == 0) {
            throw new InvalidOperationError('Unable to pop element from Empty Deque');
        }
        return this.list.popBack();
    }
    popFront() {
        if (this.list.size == 0) {
            throw new InvalidOperationError('Unable to pop element from Empty Deque');
        }
        return this.list.popFront();
    }
}
export default Deque;
